.. _pdp-citations:

==============
Citing ``pdp``
==============

Please cite one or more of the following manuscripts in your work, if you have found ``pdp`` useful:

    Pritchard L, Holden NJ, Bielaszewska M, Karch H, Toth IK (2012) Alignment-Free Design of Highly Discriminatory Diagnostic Primer Sets for Escherichia coli O104:H4 Outbreak Strains. *PLOS ONE* **7** (4): e34498. doi:10.1371/journal.pone.0034498

 .. code-block:: latex

    @article{10.1371/journal.pone.0034498,
        author = {Pritchard, Leighton AND Holden, Nicola J. AND Bielaszewska, Martina AND Karch, Helge AND Toth, Ian K.},
        journal = {PLOS ONE},
        publisher = {Public Library of Science},
        title = {Alignment-Free Design of Highly Discriminatory Diagnostic Primer Sets for Escherichia coli O104:H4 Outbreak Strains},
        year = {2012},
        month = {04},
        volume = {7},
        url = {https://doi.org/10.1371/journal.pone.0034498},
        pages = {1-8},
        abstract = {Background An Escherichia coli O104:H4 outbreak in Germany in summer 2011 caused 53 deaths, over 4000 individual infections across Europe, and considerable economic, social and political impact. This outbreak was the first in a position to exploit rapid, benchtop high-throughput sequencing (HTS) technologies and crowdsourced data analysis early in its investigation, establishing a new paradigm for rapid response to disease threats. We describe a novel strategy for design of diagnostic PCR primers that exploited this rapid draft bacterial genome sequencing to distinguish between E. coli O104:H4 outbreak isolates and other pathogenic E. coli isolates, including the historical hæmolytic uræmic syndrome (HUSEC) E. coli HUSEC041 O104:H4 strain, which possesses the same serotype as the outbreak isolates.   Methodology/Principal Findings Primers were designed using a novel alignment-free strategy against eleven draft whole genome assemblies of E. coli O104:H4 German outbreak isolates from the E. coli O104:H4 Genome Analysis Crowd-Sourcing Consortium website, and a negative sequence set containing 69 E. coli chromosome and plasmid sequences from public databases. Validation in vitro against 21 ‘positive’ E. coli O104:H4 outbreak and 32 ‘negative’ non-outbreak EHEC isolates indicated that individual primer sets exhibited 100% sensitivity for outbreak isolates, with false positive rates of between 9% and 22%. A minimal combination of two primers discriminated between outbreak and non-outbreak E. coli isolates with 100% sensitivity and 100% specificity.   Conclusions/Significance Draft genomes of isolates of disease outbreak bacteria enable high throughput primer design and enhanced diagnostic performance in comparison to traditional molecular assays. Future outbreak investigations will be able to harness HTS rapidly to generate draft genome sequences and diagnostic primer sets, greatly facilitating epidemiology and clinical diagnostics. We expect that high throughput primer design strategies will enable faster, more precise responses to future disease outbreaks of bacterial origin, and help to mitigate their societal impact.},
        number = {4},
        doi = {10.1371/journal.pone.0034498}
    } 

    Pritchard, L. , Humphris, S. , Saddler, G. S., Parkinson, N. M., Bertrand, V. , Elphinstone, J. G. and Toth, I. K. (2013), Detection of phytopathogens of the genus Dickeya using a PCR primer prediction pipeline for draft bacterial genome sequences. *Plant Pathology*, **62** : 587-596. doi:10.1111/j.1365-3059.2012.02678.x

.. code-block:: latex

    @article{doi:10.1111/j.1365-3059.2012.02678.x,
        author = {Pritchard, L. and Humphris, S. and Saddler, G. S. and Parkinson, N. M. and Bertrand, V. and Elphinstone, J. G. and Toth, I. K.},
        title = {Detection of phytopathogens of the genus Dickeya using a PCR primer prediction pipeline for draft bacterial genome sequences},
        journal = {Plant Pathology},
        volume = {62},
        number = {3},
        pages = {587-596},
        keywords = {diagnostics, Erwinia spp., Pectobacterium spp., potato, real-time PCR},
        doi = {10.1111/j.1365-3059.2012.02678.x},
        url = {https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1365-3059.2012.02678.x},
        eprint = {https://onlinelibrary.wiley.com/doi/pdf/10.1111/j.1365-3059.2012.02678.x},
        abstract = {This study used a novel computational pipeline to exploit draft bacterial genome sequences in order to predict, automatically and rapidly, PCR primer sets for Dickeya spp. that were unbiased in terms of diagnostic gene choice. This pipeline was applied to 16 draft and four complete Dickeya genome sequences to generate >700 primer sets predicted to discriminate between Dickeya at the species level. Predicted diagnostic primer sets for both D. dianthicola (DIA-A and DIA-B) and ‘D. solani’ (SOL-C and SOL-D) were validated against a panel of 70 Dickeya reference strains, representative of the known diversity of this genus, to confirm primer specificity. The classification of the four previously sequenced strains was re-examined and evidence of possible misclassification of three of these strains is presented.}
    }

.. _10.1371/journal.pone.0034498: https://dx.doi.org/10.1371/journal.pone.0034498
.. _10.1111/j.1365-3059.2012.02678.x: https://dx.doi.org/10.1111/j.1365-3059.2012.02678.x